### PR TITLE
Convert an ExecutableStage to a ProcessBundleDescriptor

### PR DESCRIPTION
**Pull request by tgroh**
This requires:
* Adding the RemoteGrpc(Read/Write)Nodes to the input and output `PCollections`
* Adding the ability to construct the wire coder from a PCollection and Components
* Adding the ability to instantiate unknown coders as Byte Array Coders

There are also associated improvements to some Control Client classes.

This is built on top of #4717 and #4695, and will be rebased as those changes go in.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by tgroh** [comment]
FYI: @bsidhom

**Comment by tgroh** [comment]
R: @bsidhom

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                        FunctionSpec.newBuilder()
+                            .setUrn(validateCommonUrn("beam:coder:windowed_value:v1"))))
+            .build();
+    // Add the original WindowedValue<T, W> coder to the components;
+    String windowedValueId =
+        uniquifyId(String.format("fn/wire/%s", pCollectionNode.getId()), usedIds);
+    return LengthPrefixUnknownCoders.forCoder(
+        windowedValueId,
+        components.toBuilder().putCoders(windowedValueId, windowedValueCoder).build(),
+        false);
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java]: `false` -> `false /* for sdk */`

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        "Unexpected Deserialized %s type, expected %s, got %s",
+        RunnerApi.Coder.class.getSimpleName(),
+        FullWindowedValueCoder.class.getSimpleName(),
+        javaCoder.getClass());
+    return (Coder<WindowedValue<?>>) javaCoder;
+  }
+
+  /** */
+  @AutoValue
+  public abstract static class SimpleProcessBundleDescriptor {
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java]: ExecutableProcessBundleDescriptor?

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            .putAllPcollections(components.getPcollectionsMap())
+            .putAllWindowingStrategies(components.getWindowingStrategiesMap())
+            .putAllTransforms(
+                stage
+                    .getTransforms()
+                    .stream()
+                    .collect(
+                        Collectors.toMap(PTransformNode::getId, PTransformNode::getTransform)));
+
+    RemoteInputDestination<WindowedValue<?>> inputDestination =
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java]: Using RemoteInputDestination works but creates only half the story since the SdkHarnessClient also defines how to get output and you can't satisfy that right now since you don't have a gRPC service.

Maybe we should extract RemoteInputDestination as a different type which can be shared between this and the SdkHarnessClient so its not coupled there.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.fn.data.RemoteGrpcPortRead;
+import org.apache.beam.sdk.fn.data.RemoteGrpcPortWrite;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.util.WindowedValue.FullWindowedValueCoder;
+
+/** Utility methods for creating {@link ProcessBundleDescriptor} instances. */
+// TODO: Rename to ExecutableStages?
+public class ProcessBundleDescriptors {
+  public static SimpleProcessBundleDescriptor fromExecutableStage(
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java]: Why wouldn't we do this as part of the ExecutableStage conversion and make the ProcessBundleDescriptor the executable stage payload?

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+  private transient ExecutorService serverExecutor;
+  private transient ExecutorService sdkHarnessExecutor;
+
+  @Before
+  public void setup() throws Exception {
+    // Setup execution-time servers
+    ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).build();
+    serverExecutor = Executors.newCachedThreadPool(threadFactory);
+    InProcessServerFactory serverFactory = InProcessServerFactory.create();
```
Comment (lukecwik) [inline runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java]: It would be valuable for test reasons to make a JUnit rule which spins up an SDK harness and tears it down as its useful in many tests. Separate out as a different PR.

Comment (tgroh) [reply runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/RemoteExecutionTest.java]: https://github.com/apache/beam/pull/4825

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
run java gradle precommit

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslation.java [hunk 6]
```diff
@@ -156,4 +157,36 @@
...
   }
+
+  /**
+   * Convert a {@link RunnerApi.Coder} into a java {@link Coder}. If there is a registered {@link
+   * CoderTranslator} for the {@link RunnerApi.Coder}, deserialize the coder using that {@link
+   * CoderTranslator}. Otherwise, return a {@link ByteArrayCoder}.
+   *
+   * <p>Component {@link RunnerApi.Coder coders} are recursively deserialized with this method.
+   *
+   * <p>Length prefix coders are not added or removed by this method.
```
[the remark this answers is not in the document]
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslation.java]: Agreed.

**Comment by lukecwik** [comment]
Please push up the changes that you said were Done and self merge.

**Comment by tgroh** [comment]
run java precommit
